src/app/Rules/MatchExpressionEvaluator.tsx

```diff
@@ -0,0 +1,141 @@
+ */
+import * as React from 'react';
+import { CodeBlock, CodeBlockCode, Label, Stack, StackItem, Text, ValidatedOptions } from '@patternfly/react-core';
+import { ServiceContext } from '@app/Shared/Services/Services';
+import { useSubscriptions } from '@app/utils/useSubscriptions';
+import { Target } from '@app/Shared/Services/Target.service';
+import { TargetSelect } from '@app/TargetSelect/TargetSelect';
+import { NoTargetSelected } from '@app/TargetView/NoTargetSelected';
+import {CheckCircleIcon, InfoCircleIcon, QuestionCircleIcon, WarningTriangleIcon} from '@patternfly/react-icons';
+
+export interface MatchExpressionEvaluatorProps {
+  matchExpression?: string;
+}
+
+export const MatchExpressionEvaluator: React.FunctionComponent<MatchExpressionEvaluatorProps> = (props) => {
+  const context = React.useContext(ServiceContext);
+  const addSubscription = useSubscriptions();
+  const [target, setTarget] = React.useState(undefined as Target | undefined);
+  const [valid, setValid] = React.useState(ValidatedOptions.default);
+
+  React.useEffect(() => {
+    addSubscription(
+      context.target.target().subscribe(setTarget)
+    );
+  }, [context, context.target]);
+
+  React.useEffect(() => {
+    if (!props.matchExpression || !target?.connectUrl) {
+      setValid(ValidatedOptions.default);
+      return;
+    }
+    try {
+      const f = new Function('target', `return ${props.matchExpression}`);
+      const res = f(target);
+      if (typeof res === 'boolean') {
+        setValid(res ? ValidatedOptions.success : ValidatedOptions.warning);
+        return;
+      }
+      setValid(ValidatedOptions.error);
+      return;
+    } catch (err) {
+      setValid(ValidatedOptions.error);
+      return;
+    }
+  }, [target, props.matchExpression]);
+
+  const statusLabel = React.useMemo(() => {
+    switch (valid) {
+      case ValidatedOptions.success:
+        return (<Label color="green" icon={<CheckCircleIcon />}>Expression Matches Selected Target</Label>);
+      case ValidatedOptions.warning:
+        return (<Label color="orange" icon={<QuestionCircleIcon />}>Expression Valid, No Match</Label>);
+      case ValidatedOptions.error:
+        return (<Label color="red" icon={<WarningTriangleIcon />}>Invalid Match Expression</Label>);
+      default:
+        return (<Label color="grey" icon={<InfoCircleIcon />}>No Match Expression</Label>);
+    }
+  }, [valid]);
+
+  const exampleExpression = React.useMemo(() => {
+    let body: string;
+    if (!target || !target?.alias || !target?.connectUrl) {
+      body = 'true';
+    } else {
+      body = `target.alias == '${target?.alias}' || target.annotations.cryostat['PORT'] == ${target?.annotations?.cryostat['PORT']}`;
+    }
+    return (<>
+      <CodeBlock>
+        <CodeBlockCode>
+          { JSON.stringify(body, null, 2) }
+        </CodeBlockCode>
+      </CodeBlock>
+    </>);
+  },[target]);
+
+  return (<>
+    <Stack hasGutter>
+      <StackItem>
+        <TargetSelect/>
```

Comment:
In the Create Recording form, the `TargetSelect` is part of the form submission. For the Automated Rules page, the`TargetSelect` and the generated example text isn't part of the form submission, even though you have to enter a match expression in the form to see the `statusLabel` validation. It would be nice if there was a clearer indication that the TargetSelect/examples are not part of the form submission, and that the validation `statusLabel` applies to the MatchExpression field on the left.

Also, to verify that my match expression applies to the right targets, I need to select each target in the `TargetSelect` dropdown and check each `statusLabel`. I think this is fine for this PR, but I was wondering what ideas you had to improve the target selection for multiple targets in the future?

Reply:
> In the Create Recording form, the TargetSelect is part of the form submission. For the Automated Rules page, theTargetSelect and the generated example text isn't part of the form submission, even though you have to enter a match expression in the form to see the statusLabel validation. It would be nice if there was a clearer indication that the TargetSelect/examples are not part of the form submission, and that the validation statusLabel applies to the MatchExpression field on the left.

Makes sense. Do you have any ideas on how to give a better conceptual link between the left side matchExpression form field and the right side component? Would adding a header/title to the card on the right help? Or perhaps that right side component should be hidden by default, and should only appear if some helper button/icon next to the matchExpression form field is clicked?

> Also, to verify that my match expression applies to the right targets, I need to select each target in the TargetSelect dropdown and check each statusLabel. I think this is fine for this PR, but I was wondering what ideas you had to improve the target selection for multiple targets in the future?

Definitely. This is something I want to find a better solution for, but I need to think about how to scale this across a large number of targets and have it work visually. Right now I'm thinking of having an entirely separate view, something like a Match Expression Wizard, that has some visualization like a list of target aliases with a checkbox to indicate whether the given expression matches the target. The list items would be expandable to reveal the whole JSON structure of the target like we see in this PR, so that the user can look at the details of all of the target's fields and use that to craft their expression. Collapsing all of the list items so that the view is just the list of checkboxes would give the user a way to evaluate whether the expression is matching all of the expected targets in a compact way. But I'm not completely happy with this solution either because it still means there will be a potentially very long list of targets and a lot of checkboxes to look at, so it will scale up to a point and then become very cumbersome again.

This separate view would probably be something like a modal that can be summoned from within the Create Rule form, and some other similar view could also be on the main Rules page so that users can immediately relate an existing Rule to the targets it applies to.

Reply:
The header and the validation within the match expression text field looks much better. I was going to suggest something similar.

I'd rather keep the match expression evaluator visible at all times. There's enough space on the page to do so, and keeping it as a separate item on the side helps to indicate that it's not part of the form.